### ISSUE TITLE
Seemingly incorrect training behavior in `lolcats_llama_window_sw`

### ISSUE THREAD

**Issue by agostinv** [HazyResearch/lolcats#9]
Half-seeking clarification and half-raising a potential issue, here (if I was certain I'd just make a PR, but new to this codebase). It looks like `lolcats_llama_window_sw` doesn't correctly distill, although it should work during fine-tuning and evaluation (of course). The following lines from `linear_attention`

https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_attention.py#L322-L334

don't have analogous lines in `lolcats_llama_window_sw` for the classical, softmax attention necessary for distillation:

https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw_linear.py#L230-L243

While most users should use the ThunderKittens implementation anyways, it would be ideal to unify these implementations.

**Comment by agostinv** [comment]
@mzio

**Comment by mzio** [comment]
hey @agostinv thanks for bringing this up!

Sorry for the confusing code upload. If you're running the sliding window implementation from [`configs/model/distill_llama3_1_8b_lk_smd_wsw64_fd64_w01.yaml`](https://github.com/HazyResearch/lolcats/blob/main/configs/model/distill_llama3_1_8b_lk_smd_wsw64_fd64_w01.yaml), e.g. using
```bash
--model_config distill_llama3_1_8b_lk_smd_wsw64_fd64_w01
```
then from 
```yaml
attention:
  attention_type: lolcats_llama_window_sw
  state_chunk_len: 1024
  window_size: 64
  affine_attention_factors: false
  init_window_factor: -2.1972245773362196
```

it should pull up the implementation in [`lolcats/src/model/linear_attention/linear_window_attention_sw.py`](https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw.py#L142-L156)

https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw.py#L142-L156

which mirrors the `train_attention` logic in `linear_attention.py`

The implementations above are meant to be "didactic" references that make most clear what we are doing (but as you might see involve computing the n^2 attention weights for both the linear attentions and softmax attentions, if n is input sequence length).  

---

However, because we just need the layer and head-wise attention outputs, we can also do this training in O(n) memory. That's what your referenced `linear_window_attention_sw_linear.py` does, which is meant to be called with `src/trainer/distill_attention_mse_linear.py` (note the `_linear.py`; sorry we should upload an experiment config for this, e.g., a variation on `configs/experiment/distill_alpaca_clean_xent0_mse1000_lr1e-2.yaml` that replaces
```yaml
...
trainer:  # HuggingFace Trainer-like arguments  
  name: distill_attention_xent_mse
...
```
with 
```yaml
...
trainer:  # HuggingFace Trainer-like arguments  
  name: distill_attention_mse_linear
...
```

Here for distillation we first compute all the softmax attention outputs with FlashAttention (end-to-end as a single forward pass)

**Part of Training Loop:**
https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/trainer/distill_attention_mse_linear.py#L45-L63

**Part of Model Forward Pass:**
https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw_linear.py#L214-L228

Before then computing all the linear attention outputs with the sliding window setup:  

**Part of Training Loop:**
https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/trainer/distill_attention_mse_linear.py#L72-L83

**Part of Model Forward Pass:**
https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw_linear.py#L231-L243

where `self.linear_attention` is earlier set to reference:
https://github.com/HazyResearch/lolcats/blob/d2d47d1b616ef90efd0bd8526e747bdfcba46dd0/src/model/linear_attention/linear_window_attention_sw_linear.py#L133-L157

**Comment by agostinv** [comment]
Hey @mzio, thanks for the speedy clarification and guiding me towards what I was looking for, really appreciate it! Sorry for the confusion haha

Closing as my question was answered.
